rclcpp/src/rclcpp/subscription_base.cpp

```diff
@@ -288,3 +306,83 @@ SubscriptionBase::exchange_in_use_by_wait_set_state(
   }
   throw std::runtime_error("given pointer_to_subscription_part does not match any part");
 }
+
+bool
+SubscriptionBase::is_cft_supported() const
+{
+  return rcl_subscription_is_cft_supported(subscription_handle_.get());
+}
+
+void
+SubscriptionBase::set_cft_expression_parameters(
+  const std::string & filter_expression,
+  const std::vector<std::string> & expression_parameters)
+{
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  rcutils_string_array_t parameters;
+  parameters = rcutils_get_zero_initialized_string_array();
+  rcutils_ret_t rcutils_ret = rcutils_string_array_init(
+    &parameters, expression_parameters.size(), &allocator);
+  if (RCUTILS_RET_OK != rcutils_ret) {
+    rclcpp::exceptions::throw_from_rcl_error(
+      RCL_RET_ERROR,
+      "failed to initialize string array for expression parameters");
+  }
+  RCLCPP_SCOPE_EXIT(
+  {
+    rcutils_ret_t rcutils_ret = rcutils_string_array_fini(&parameters);
+    if (RCUTILS_RET_OK != rcutils_ret) {
+      RCLCPP_ERROR(
+        rclcpp::get_node_logger(node_handle_.get()).get_child("rclcpp"),
+        "failed to finalize parameters: %s",
+        rcl_get_error_string().str);
+      rcl_reset_error();
+    }
+  });
+
+  for (size_t i = 0; i < expression_parameters.size(); ++i) {
+    parameters.data[i] = rcutils_strdup(expression_parameters[i].c_str(), allocator);
+  }
+  rcl_ret_t ret = rcl_subscription_set_cft_expression_parameters(
+    subscription_handle_.get(),
+    filter_expression.c_str(),
+    &parameters);
+
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret, "failed to set cft expression parameters");
+  }
+}
+
+void
+SubscriptionBase::get_cft_expression_parameters(
+  std::string & filter_expression,
+  std::vector<std::string> & expression_parameters) const
+{
+  char * expression = NULL;
+  rcutils_string_array_t parameters;
+  parameters = rcutils_get_zero_initialized_string_array();
+  rcl_ret_t ret = rcl_subscription_get_cft_expression_parameters(
+    subscription_handle_.get(),
+    &expression,
+    &parameters);
+
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret, "failed to get cft expression parameters");
+  }
+
+  filter_expression = expression;
+  for (size_t i = 0; i < parameters.size; ++i) {
+    expression_parameters.push_back(parameters.data[i]);
```

Comment:
emplace_back() is better?

Reply:
The type of `parameters.data[i]` is char*, `emplace_back` can't take the owner of it, and I don't need the return value by `emplace_back` if C++14 standard is effective.

Actually, push_back already provides `push_back(value_type&& __x)` from C++11 which will call emplace_back directly if the value can be moved.

Reply:
> The type of parameters.data[i] is char*, emplace_back can't take the owner of it, and I don't need the return value by emplace_back if C++14 standard is effective.
> Actually, push_back already provides push_back(value_type&& __x) from C++11 which will call emplace_back directly if the value can be moved.

Yeah. It's my mistake.